### PR TITLE
Also support non-contiguous activation for torch._weight_int8pack_mm on CPU

### PR DESCRIPTION
**Pull request by sanchitintel**
### Problem
Non-contiguous activation for `torch._weight_int8pack_mm` is unsupported on CPU.
So, with int8 WoQ with B16 activation with torchao, for batch-size 2 & above, an assertion is hit regarding non-contiguous A being unsupported. Such an issue was encountered with LLaMA models.

### Solution
Also support non-contiguous activation for `torch._weight_int8pack_mm`, so long as it's contiguous on the last dimension & remove the assertion that requires contiguous activation.

### Alternative solutions considered
Could modify LLaMA model in transformers library to call `contiguous` after obtaining the final hidden state, just before computing logits with the LM head. However, [it](https://github.com/huggingface/transformers/pull/36078) might cause some regression for other users of that code. 

Another aspect to this issue is - is latency always lower if we make an activation tensor contiguous before linear or `torch._weight_int8pack_mm` is called on CPU? I guess we need some data-points to analyze this part, although I think the performance should be good enough with this patch, since the first cache lines of rows of A are being explicitly prefetched in the existing code (and it also avoids copy, which a `contiguous` call would do).

cc @jgong5 @mingfeima @XiaobingSuper @ashokei @jingxu10

### CODE REVIEW

**Comment by sanchitintel** [comment]
Hi @mingfeima,

Seeking your advice regarding `torch._weight_int8pack_mm` - currently it doesn't support non-contiguous activations, but in LLaMA, the activation to LM head is non-contiguous with batch size >= 2, so an assert is being hit.

Please advise if you would recommend supporting non-contiguous activations in `torch._weight_int8pack_mm` to make the functionality work, or if you support adding a call to `contiguous()` for the activation of LM head in LLaMA code in transformers library, before calling linear.

I mean to ask, should we gather data on which one would be better for various input shapes, and then decide whether to modify ATen code or LLaMA model in transformers library  (perhaps only for CPU)?

Thanks!

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 1]
```diff
@@ -3572,11 +3572,10 @@ Tensor _weight_int8pack_mm_cpu(
 
   TORCH_CHECK(A.dtype() == kBFloat16 || A.dtype() == kHalf || A.dtype() == kFloat,
       __func__, " : expect A to be either 32-bit or 16-bit float tensor.");
-  TORCH_CHECK(A.is_contiguous(),
-      __func__, " : expect A to be contiguous.");
   TORCH_CHECK(A.dim() == 2,
       __func__, " : expect A to be 2D tensor.");
-
+  TORCH_CHECK(A.stride(1) == 1,
+      __func__, " : A must be row-major even if it's non-contiguous");
```
Comment (mingfeima) [inline aten/src/ATen/native/LinearAlgebra.cpp:3578]: change the error message to "A must be contiguous on last dimension."

**Review by mingfeima — APPROVED** [review]
don't need to benchmark this, won't affect the performance. just change the error message a little bit.

**Review by leslie-fang-intel — APPROVED** [review]

**Comment by sanchitintel** [comment]
Hi @malfet, can you please help review this PR? It requires approval from a core reviewer/maintainer for landing. 

Thanks!

**Comment by malfet** [comment]
> Hi @malfet, can you please help review this PR? It requires approval from a core reviewer/maintainer for landing. 
> 
> 
> 
> Thanks!

Hmm, this change looks BC breaking to me(at quick glance, afk atm). Any reason making it incompatible with contiguous A?

**Review by malfet — APPROVED** [review]
LGTM, thank you for the explanation
